[AIRFLOW-1273] Add Google Cloud ML version and model operators

N3da opened the pull request:
### JIRA
- [ https://issues.apache.org/jira/browse/AIRFLOW-1273]

### Description
Includes the following changes:
- gcp_cloudml_hooks.py to work Google Cloud ML engine Rest APIs for model and versions (docs:
https://cloud.google.com/ml-engine/reference/rest/) 
- cloudml_operator.py which contains `CloudMLVersionOperator` and `CloudMLModelOperator`s
- unit tests for the added hooks.
- Adding a default value for `google_cloud_default` connection.

### Tests
- Added unit tests for all added methods in gcp_cloudml_hooks.py.
[Link to Travis: 
https://travis-ci.org/N3da/incubator-airflow/builds/244668457]

bolkedebruin commented on airflow/contrib/hooks/gcp_cloudml_hook.py:
```diff
@@ -0,0 +1,165 @@
+from oauth2client.client import GoogleCredentials
+from airflow.contrib.hooks.gcp_api_base_hook import GoogleCloudBaseHook
+
+logging.getLogger('GoogleCloudML').setLevel(logging.INFO)
```
You can import the logging level from settings.py (settings.LOGGING_LEVEL)

N3da replied:
Thanks. Done.

bolkedebruin commented on airflow/contrib/hooks/gcp_cloudml_hook.py:
```diff
@@ -0,0 +1,165 @@
+                time.sleep((2**i) + (random.randint(0, 1000) / 1000))
+        except errors.HttpError as e:
+            if e.resp.status != 429:
+                logging.info('Something went wrong. Not retrying: %s', e)
```
please use .format

N3da replied:
Done.

bolkedebruin commented on airflow/contrib/hooks/gcp_cloudml_hook.py:
```diff
@@ -0,0 +1,165 @@
+
+        try:
+            response = request.execute()
+            logging.info('Successfully set version: %s to default', response)
```
please use .format

N3da replied:
Done.

bolkedebruin commented on airflow/contrib/hooks/gcp_cloudml_hook.py:
```diff
@@ -0,0 +1,165 @@
+            logging.info('Successfully set version: %s to default', response)
+            return response
+        except errors.HttpError as e:
+            logging.error('Something went wrong: %s', e)
```
idem

N3da replied:
Done.

bolkedebruin commented on airflow/contrib/hooks/gcp_cloudml_hook.py:
```diff
@@ -0,0 +1,165 @@
+            return request.execute()
+        except errors.HttpError as e:
+            if e.resp.status == 404:
+                logging.error('Model was not found: %s', e)
```
idem

N3da replied:
Done.

bolkedebruin commented on airflow/contrib/operators/cloudml_operator.py:
```diff
@@ -0,0 +1,101 @@
+from airflow.operators import BaseOperator
+from airflow.utils.decorators import apply_defaults
+
+logging.getLogger('GoogleCloudML').setLevel(logging.INFO)
```
settings.LOGGING_LEVEL

N3da replied:
Done.

N3da committed:
Add Google Cloud ML version and model operators

Includes Google Cloud ML hooks for version and model operations,
and their unit tests.

https://issues.apache.org/jira/browse/AIRFLOW-1273

criccomini commented on airflow/contrib/hooks/gcp_cloudml_hook.py:
```diff
@@ -0,0 +1,166 @@
+        return build('ml', 'v1', credentials=credentials)
+
+    def create_version(self, project_name, model_name, version_spec):
+        """Creates the Version on Cloud ML.
```
Nit: newline after """

N3da replied:
Done.

criccomini commented on airflow/contrib/operators/cloudml_operator.py:
```diff
@@ -0,0 +1,103 @@
+logging.getLogger('GoogleCloudML').setLevel(settings.LOGGING_LEVEL)
+
+
+class CloudMLVersionOperator(BaseOperator):
```
Some pydocs here would be helpful. The pydocs from these classes gets rolled into the public documentation automatically, so being extra verbose is actually good for the community.

N3da replied:
Done.

criccomini commented on airflow/contrib/operators/cloudml_operator.py:
```diff
@@ -0,0 +1,103 @@
+            raise ValueError('Unknown operation: {}'.format(self._operation))
+
+
+class CloudMLModelOperator(BaseOperator):
```
Some pydocs here would be helpful. The pydocs from these classes gets rolled into the public documentation automatically, so being extra verbose is actually good for the community.

N3da replied:
Done.

criccomini commented on airflow/utils/db.py:
```diff
@@ -130,6 +130,10 @@ def initdb():
             schema='hive', port=3400))
     merge_conn(
         models.Connection(
+            conn_id='google_cloud_default', conn_type='google_cloud_platform',
```
Not sure how useful this is. Was there any particular reason you added it? If you want to keep it, having an example `extra={}` defined (as with beeline above) could be of some use, though it'd point to some service account json file that doesn't actually exist.

N3da replied:
Ah. This was needed for the unit tests which start from the empty db, and it would throw when initializing the gcp_cloudml_hook without having a value for `google_cloud_default` (and the tests would be skipped). Not sure if the `extra` args would be useful though.

criccomini commented on tests/contrib/hooks/test_gcp_cloudml_hook.py:
```diff
@@ -0,0 +1,254 @@
+class _TestCloudMLHook(object):
+
+    def __init__(self, test_cls, responses, expected_requests):
+        """Init method.
```
Nit: newline after """

N3da replied:
done.

criccomini commented on airflow/contrib/hooks/gcp_cloudml_hook.py:
```diff
@@ -0,0 +1,166 @@
+logging.getLogger('GoogleCloudML').setLevel(settings.LOGGING_LEVEL)
+
+
+def _retry_with_exponential_delay(request, max_n, is_done_func, is_error_func):
```
One thing to consider: Airflow operators, themselves, have a retry setting that developers can use in the case of a failure. If you simply allow failures to propagate up, the task will fail, and the developer can decide whether the task should retry. It might be cleaner/easier to remove this retry logic, and just allow failures to go up the stack to Airflow, where config can dictate how to handle failures.

It is nice to have slightly more sophisticated logic, as you do, to handle retry HTTP error codes, and fail outright on others, so I could see an argument for keeping this. Just something to consider.

N3da replied:
Thanks Chris. The purpose of retry here is slightly different. Some of our http calls (for example for creating a `Version`, return a long running `[Operation](https://cloud.google.com/ml-engine/reference/rest/Shared.Types/ListOperationsResponse#Operation)` which the user needs to keep polling to determine when it's done. This function is to avoid polling every `n` seconds which might hit user's quota restrictions. In the case of errors we mostly don't want to add extra retry logic on top of existing generic logic (except for the `429` status case which can be retried).

N3da replied:
On second thought renamed it to `_poll_...` instead of `_retry_...` to make it more clear.
